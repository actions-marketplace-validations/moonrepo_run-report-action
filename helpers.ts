import type { Action, ActionStatus, Duration, RunReport } from '@moonrepo/types';

export function getCommitInfo() {
	const sha = process.env.GITHUB_SHA;
	const server = process.env.GITHUB_SERVER_URL ?? 'https://github.com';
	const repo = process.env.GITHUB_REPOSITORY;

	if (!sha || !repo) {
		return null;
	}

	const pr = /refs\/pull\/(\d+)\//g.exec(process.env.GITHUB_REF!);

	if (pr?.[1]) {
		return {
			sha,
			url: `${server}/${repo}/pull/${pr[1]}/commits/${sha}`,
		};
	}

	return {
		sha,
		url: `${server}/${repo}/commit/${sha}`,
	};
}

export function getIconForStatus(status: ActionStatus): string {
	switch (status) {
		case 'cached':
			return '🟪';
		case 'failed':
		case 'failed-and-abort':
			return '🟥';
		case 'invalid':
			return '🟨';
		case 'passed':
			return '🟩';
		default:
			return '⬛️';
	}
}

export function hasFailed(status: ActionStatus): boolean {
	return status === 'failed' || status === 'failed-and-abort';
}

export function hasPassed(status: ActionStatus): boolean {
	return status === 'passed' || status === 'cached';
}

export function isFlaky(action: Action): boolean {
	if (!action.attempts || action.attempts.length === 0) {
		return false;
	}

	const someAttemptsFailed = action.attempts.some((attempt) => hasFailed(attempt.status));

	return hasPassed(action.status) && someAttemptsFailed;
}

export function formatTime(mins: number, secs: number, millis: number): string {
	if (mins === 0 && secs === 0 && millis === 0) {
		return '0s';
	}

	const value: string[] = [];

	if (mins > 0) {
		value.push(`${mins}m`);
	}

	if (secs > 0) {
		value.push(`${secs}s`);
	}

	if (millis > 0) {
		let ms = millis.toFixed(1);

		if (ms.endsWith('.0')) {
			ms = ms.slice(0, -2);
		}

		value.push(`${ms}ms`);
	}

	return value.join(', ');
}

export function formatDuration(duration: Duration | null): string {
	if (!duration) {
		return '--';
	}

	if (duration.secs === 0 && duration.nanos === 0) {
		return '0s';
	}

	let mins = 0;
	let { secs } = duration;
	const millis = duration.nanos / 1_000_000;

	while (secs > 60) {
		mins += 1;
		secs -= 60;
	}

	return formatTime(mins, secs, millis);
}

export function calculateTotalTime(report: RunReport): string {
	let mins = 0;
	let secs = 0;
	let millis = 0;

	report.actions.forEach((action) => {
		if (action.duration) {
			secs += action.duration.secs;
			millis += action.duration.nanos / 1_000_000;
		}
	});

	while (millis > 1000) {
		secs += 1;
		millis -= 1000;
	}

	while (secs > 60) {
		mins += 1;
		secs -= 60;
	}

	return formatTime(mins, secs, 0);
}

export function formatReportToMarkdown(report: RunReport): string {
	const markdown = [
		'### Run report',
		'|     | Action | Time | Status |    |',
		'| :-: | :----- | ---: | :----- | :- |',
	];

	report.actions.forEach((action) => {
		const comments: string[] = [];

		if (isFlaky(action)) {
			comments.push('**FLAKY**');
		}

		if (action.attempts && action.attempts.length > 1) {
			comments.push(`${action.attempts.length} attempts`);
		}

		if (hasFailed(action.status) && action.error) {
			comments.push(action.error);
		}

		markdown.push(
			`| ${getIconForStatus(action.status)} | \`${action.label}\` | ${formatDuration(
				action.duration,
			)} | ${action.status} | ${comments.join(', ')} |`,
		);
	});

	markdown.push(`| | | ${calculateTotalTime(report)} | | |`);

	if (report.context.touchedFiles.length > 0) {
		markdown.push('', '### Touched files', '<details><summary>View files list</summary><div>\n');

		report.context.touchedFiles.forEach((file) => {
			markdown.push(`- ${file}`);
		});

		markdown.push('\n</div></details>');
	}

	const commit = getCommitInfo();

	if (commit) {
		markdown.push(
			'',
			`<small>Run report for commit [${commit.sha.slice(0, 7)}](${commit.url})</small>`,
		);
	}

	return markdown.join('\n');
}
