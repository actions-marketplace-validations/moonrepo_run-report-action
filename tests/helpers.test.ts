import * as github from '@actions/github';
import type { RunReport } from '@moonrepo/types';
import { formatReportToMarkdown, sortReport } from '../helpers';

jest.mock('@actions/github', () => ({
	context: { payload: {} },
}));

describe('formatReportToMarkdown()', () => {
	it('renders a standard report', () => {
		expect(formatReportToMarkdown(require('./__fixtures__/standard.json'))).toMatchSnapshot();
	});

	it('renders a standard report with touched files', () => {
		expect(
			formatReportToMarkdown(require('./__fixtures__/standard-touched-files.json')),
		).toMatchSnapshot();
	});

	it('renders durations', () => {
		expect(formatReportToMarkdown(require('./__fixtures__/durations.json'))).toMatchSnapshot();
	});

	it('renders attempts and flakiness', () => {
		expect(formatReportToMarkdown(require('./__fixtures__/flakiness.json'))).toMatchSnapshot();
	});

	it('renders errors', () => {
		expect(formatReportToMarkdown(require('./__fixtures__/errors.json'))).toMatchSnapshot();
	});

	it('renders with git commit', () => {
		Object.assign(github.context, {
			sha: '59719f967ddcf585da9bc7ba8730dcd2865cbdfa',
			payload: {
				pull_request: {
					number: '123',
				},
			},
		});

		process.env.GITHUB_REPOSITORY = 'moonrepo/moon';

		expect(formatReportToMarkdown(require('./__fixtures__/standard.json'))).toMatchSnapshot();

		delete process.env.GITHUB_REPOSITORY;
	});
});

describe('sortReport()', () => {
	it('sorts by time', () => {
		const report = require('./__fixtures__/standard.json') as RunReport;

		sortReport(report, 'time', 'desc');

		expect(formatReportToMarkdown(report)).toMatchSnapshot();
	});

	it('sorts by label', () => {
		const report = require('./__fixtures__/standard.json') as RunReport;

		sortReport(report, 'label', 'asc');

		expect(formatReportToMarkdown(report)).toMatchSnapshot();
	});
});
