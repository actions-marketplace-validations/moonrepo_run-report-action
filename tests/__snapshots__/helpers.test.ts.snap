// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatReportToMarkdown() renders a standard report 1`] = `
"<!-- moon-run-report -->
### Run report
|     | Action | Time | Status | Info |
| :-: | :----- | ---: | :----- | :--- |
| ⬛️ | \`SetupNodeToolchain\` | 1.4ms | skipped |  |
| ⬛️ | \`SyncNodeProject(types)\` | 3.2ms | skipped |  |
| ⬛️ | \`InstallNodeDeps\` | 3.6ms | skipped |  |
| ⬛️ | \`SyncNodeProject(runtime)\` | 3.7ms | skipped |  |
| ⬛️ | \`SyncNodeProject(website)\` | 4ms | skipped |  |
| 🟪 | \`RunTarget(types:build)\` | 310.2ms | cached |  |
| 🟩 | \`RunTarget(runtime:typecheck)\` | 2s, 380.2ms | passed |  |
| 🟩 | \`RunTarget(types:typecheck)\` | 571.6ms | passed |  |
| 🟩 | \`RunTarget(website:typecheck)\` | 4s, 431.2ms | passed |  |
| | | 7s | | |"
`;

exports[`formatReportToMarkdown() renders a standard report with touched files 1`] = `
"<!-- moon-run-report -->
### Run report
|     | Action | Time | Status | Info |
| :-: | :----- | ---: | :----- | :--- |
| ⬛️ | \`SetupNodeToolchain\` | 1.4ms | skipped |  |
| ⬛️ | \`SyncNodeProject(types)\` | 3.2ms | skipped |  |
| ⬛️ | \`InstallNodeDeps\` | 3.6ms | skipped |  |
| ⬛️ | \`SyncNodeProject(runtime)\` | 3.7ms | skipped |  |
| ⬛️ | \`SyncNodeProject(website)\` | 4ms | skipped |  |
| 🟪 | \`RunTarget(types:build)\` | 310.2ms | cached |  |
| 🟩 | \`RunTarget(runtime:typecheck)\` | 2s, 380.2ms | passed |  |
| 🟩 | \`RunTarget(types:typecheck)\` | 571.6ms | passed |  |
| 🟩 | \`RunTarget(website:typecheck)\` | 4s, 431.2ms | passed |  |
| | | 7s | | |

### Touched files
<details><summary>View files list</summary><div>

\`\`\`
packages/example/src/index.ts
apps/web/src/index.tsx
\`\`\`

</div></details>"
`;

exports[`formatReportToMarkdown() renders attempts and flakiness 1`] = `
"<!-- moon-run-report -->
### Run report
|     | Action | Time | Status | Info |
| :-: | :----- | ---: | :----- | :--- |
| 🟩 | \`RunTarget(runtime:typecheck)\` | 2s, 380.2ms | passed | **FLAKY**, 3 attempts |
| 🟩 | \`RunTarget(runtime:typecheck)\` | 2s, 380.2ms | passed | 2 attempts |
| 🟥 | \`RunTarget(runtime:typecheck)\` | 2s, 380.2ms | failed | 3 attempts |
| | | 7s | | |"
`;

exports[`formatReportToMarkdown() renders durations 1`] = `
"<!-- moon-run-report -->
### Run report
|     | Action | Time | Status | Info |
| :-: | :----- | ---: | :----- | :--- |
| 🟪 | \`RunTarget(types:build)\` | 0s | cached |  |
| 🟩 | \`RunTarget(runtime:typecheck)\` | 32m, 2s, 380.2ms | passed |  |
| 🟩 | \`RunTarget(types:typecheck)\` | 1m, 4s, 571.6ms | passed |  |
| 🟩 | \`RunTarget(website:typecheck)\` | 34s, 431.2ms | passed |  |
| | | 33m, 41s | | |"
`;

exports[`formatReportToMarkdown() renders errors 1`] = `
"<!-- moon-run-report -->
### Run report
|     | Action | Time | Status | Info |
| :-: | :----- | ---: | :----- | :--- |
| 🟥 | \`RunTarget(types:build)\` | 0s | failed | Something weird has happened |
| 🟩 | \`RunTarget(runtime:typecheck)\` | 32m, 2s, 380.2ms | passed |  |
| | | 32m, 2s | | |"
`;

exports[`formatReportToMarkdown() renders with git commit 1`] = `
"<!-- moon-run-report -->
### Run report for [59719f96](https://github.com/moonrepo/moon/pull/123/commits/59719f967ddcf585da9bc7ba8730dcd2865cbdfa)
|     | Action | Time | Status | Info |
| :-: | :----- | ---: | :----- | :--- |
| ⬛️ | \`SetupNodeToolchain\` | 1.4ms | skipped |  |
| ⬛️ | \`SyncNodeProject(types)\` | 3.2ms | skipped |  |
| ⬛️ | \`InstallNodeDeps\` | 3.6ms | skipped |  |
| ⬛️ | \`SyncNodeProject(runtime)\` | 3.7ms | skipped |  |
| ⬛️ | \`SyncNodeProject(website)\` | 4ms | skipped |  |
| 🟪 | \`RunTarget(types:build)\` | 310.2ms | cached |  |
| 🟩 | \`RunTarget(runtime:typecheck)\` | 2s, 380.2ms | passed |  |
| 🟩 | \`RunTarget(types:typecheck)\` | 571.6ms | passed |  |
| 🟩 | \`RunTarget(website:typecheck)\` | 4s, 431.2ms | passed |  |
| | | 7s | | |"
`;

exports[`sortReport() sorts by label 1`] = `
"<!-- moon-run-report -->
### Run report
|     | Action | Time | Status | Info |
| :-: | :----- | ---: | :----- | :--- |
| ⬛️ | \`InstallNodeDeps\` | 3.6ms | skipped |  |
| 🟩 | \`RunTarget(runtime:typecheck)\` | 2s, 380.2ms | passed |  |
| 🟪 | \`RunTarget(types:build)\` | 310.2ms | cached |  |
| 🟩 | \`RunTarget(types:typecheck)\` | 571.6ms | passed |  |
| 🟩 | \`RunTarget(website:typecheck)\` | 4s, 431.2ms | passed |  |
| ⬛️ | \`SetupNodeToolchain\` | 1.4ms | skipped |  |
| ⬛️ | \`SyncNodeProject(runtime)\` | 3.7ms | skipped |  |
| ⬛️ | \`SyncNodeProject(types)\` | 3.2ms | skipped |  |
| ⬛️ | \`SyncNodeProject(website)\` | 4ms | skipped |  |
| | | 7s | | |"
`;

exports[`sortReport() sorts by time 1`] = `
"<!-- moon-run-report -->
### Run report
|     | Action | Time | Status | Info |
| :-: | :----- | ---: | :----- | :--- |
| 🟩 | \`RunTarget(website:typecheck)\` | 4s, 431.2ms | passed |  |
| 🟩 | \`RunTarget(runtime:typecheck)\` | 2s, 380.2ms | passed |  |
| 🟩 | \`RunTarget(types:typecheck)\` | 571.6ms | passed |  |
| 🟪 | \`RunTarget(types:build)\` | 310.2ms | cached |  |
| ⬛️ | \`SyncNodeProject(website)\` | 4ms | skipped |  |
| ⬛️ | \`SyncNodeProject(runtime)\` | 3.7ms | skipped |  |
| ⬛️ | \`InstallNodeDeps\` | 3.6ms | skipped |  |
| ⬛️ | \`SyncNodeProject(types)\` | 3.2ms | skipped |  |
| ⬛️ | \`SetupNodeToolchain\` | 1.4ms | skipped |  |
| | | 7s | | |"
`;
